lib/websocket.go

```diff
@@ -58,14 +102,154 @@ func (h *wsHandler) wsMessageHandler(_ context.Context, e event.Event) error {
+	// abstracts over profile & key agreement
+
+	h.connsLock.Lock()
+	c, ok := h.conns[connID]
+	if !ok {
+		return fmt.Errorf("no connection for connection ID %q found", connID)
+	}
+	c.profileID = claims.Subject
+	h.connsLock.Unlock()
+
+	h.subLock.Lock()
+	defer h.subLock.Unlock()
+	h.subscriptions[claims.Subject] = connID
 	return nil
 }
+
+// unsubscribeConn remove the profileID and connID from the map of "subscribed"
+// connections
+func (h *wsHandler) unsubscribeConn(profileID string) {
+	h.subLock.Lock()
+	defer h.subLock.Unlock()
+	delete(h.subscriptions, profileID)
+}
+
+// removeConn removes the conn from the map of connections and subscriptions
+// closing the connection if needed
+func (h *wsHandler) removeConn(connID string) {
+	c, ok := h.conns[connID]
+	if !ok {
+		return
+	}
+	defer func() {
+		c.conn.Close(websocket.StatusNormalClosure, "pruning connection")
+	}()
+	if c.profileID != "" {
+		h.unsubscribeConn(c.profileID)
+	}
+	h.connsLock.Lock()
+	defer h.connsLock.Unlock()
+	delete(h.conns, connID)
+}
+
+// read listens to the given connection, handling any messages that come through
+// stops listening if it encounters any error
+func (h *wsHandler) read(id string) error {
+	ctx := context.Background()
+	msg := &message{}
+	var err error
+	wsc, ok := h.conns[id]
+	if !ok {
+		return fmt.Errorf("connection for connection ID %q not found", id)
+	}
+
+	for {
+		err = wsjson.Read(ctx, wsc.conn, msg)
+		if err != nil {
+			// all websocket methods that return w/ failure
+			// close the connection
+			h.removeConn(id)
+			return err
+		}
+		go h.handleMessage(id, msg)
+	}
+}
+
+// handleMessage handles each message based on msgType
+func (h *wsHandler) handleMessage(id string, msg *message) {
+	switch msg.Type {
+	case wsSubscribe:
+		subMsg := &subscribeMessage{}
+		err := json.Unmarshal(msg.Payload, subMsg)
+		if err != nil {
+			log.Debugf("connection %q - error unmarshaling payload for subscribe message: %s", id, err)
+		}
+		h.subscribeConn(id, subMsg.Token)
+	case wsUnsubscribe:
+		c, ok := h.conns[id]
+		if !ok {
+			log.Errorf("conn not found %q", id)
+			return
+		}
+		h.unsubscribeConn(c.profileID)
+	default:
+		log.Debug("unknown message type over websocket %s: %q", id, msg.Type)
+	}
+}
+
+// msgType is the type of message that we receive on the
+type msgType string
+
+const (
+	// wsSubscribe indicates the connection is trying to become
+	// an authenticated connection
+	// payload is a `subscribeMessage`
+	wsSubscribe = msgType("subscribe")
+	// wsUnsubscribe indicates the connection no longer wants
+	// to be authenticated
+	// payload is nil
+	wsUnsubscribe = msgType("unsubscribe")
+)
+
+// message is the expected structure of an incoming websocket message
+type message struct {
+	Type    msgType         `json:"type"`
+	Payload json.RawMessage `json:"payload"`
+}
+
+// subscribeMessage is the expected structure of an incoming "subscribe"
+// message
+type subscribeMessage struct {
+	Token string `json:"token"`
+}
```

Comment:
slightly modeled after how we do `events`, but with the understanding that we are communicating using json